### PR TITLE
Detect more Classic Assert

### LINKED ISSUE #789: NUnit.Analyzers 4.3.0 missed an Assert.Positive that caused a build error after upgrading to latest NUnit 4.2.2
NUnit.Analyzers 4.3.0 missed an `Assert.Positive` that caused a build error after upgrading to NUnit 4.2.2, from NUnit 3.13.3

That is there was no build error or warning when on NUnit 3.13.3 and building with NUnit.Analyzers 4.3.0.

Or if there was a warning, I missed it because I had tens of thousands of changes to make.

I suggest this becomes a build error for NUnit.Analyzers.

In this case, it was a double:
```
var elapsedTime1 = results.DurationInSeconds;//NOTE: elapsedTime1 is a double
Assert.Positive(elapsedTime1);
```
Which I changed to:
```
Assert.That(elapsedTime1, Is.GreaterThanOrEqualTo(0));
```

I just grabbed the excerpt from our test to make this a quick bug report.

### LINKED ISSUE #790: NUnit.Analyzers 4.3.0 missed an Assert.IsAssignableFrom that caused a build error after upgrading to latest NUnit 4.2.2
NUnit.Analyzers 4.3.0 missed an Assert.IsAssignableFrom that caused a build error after upgrading to NUnit 4.2.2, from NUnit 3.13.3.

That is there was no build error or warning when on NUnit 3.13.3 and building with NUnit.Analyzers 4.3.0.

Or if there was a warning, I missed it because I had tens of thousands of changes to make.

I suggest this becomes a build error for NUnit.Analyzers.

Fixed manually by doing this:
`Assert.IsAssignableFrom<ClassName>(objectVariable);` -> `Assert.That(objectVariable, Is.AssignableFrom(typeof(ClassName)));`

### PR DESCRIPTION
**Pull request by manfred-brands**
Fixes #789  Assert.Positive, Assert.Negative 
Fixes #790  Assert.IsAssignableFrom, Assert.IsNotAssignableFrom

### CODE REVIEW

File: src/nunit.analyzers.tests/ClassicModelAssertUsage/ClassicModelAssertUsageAnalyzerTests.cs [hunk 1]
```diff
@@ -90,6 +90,10 @@ public void VerifySupportedDiagnostics()
                     $"{AnalyzerIdentifiers.IsInstanceOfUsage} is missing.");
                 Assert.That(diagnosticIds, Contains.Item(AnalyzerIdentifiers.IsNotInstanceOfUsage),
                     $"{AnalyzerIdentifiers.IsNotInstanceOfUsage} is missing.");
+                Assert.That(diagnosticIds, Contains.Item(AnalyzerIdentifiers.PositiveUsage),
+                    $"{AnalyzerIdentifiers.PositiveUsage} is missing.");
+                Assert.That(diagnosticIds, Contains.Item(AnalyzerIdentifiers.NegativeUsage),
+                    $"{AnalyzerIdentifiers.NegativeUsage} is missing.");
             });
```
Comment (mikkelbu) [inline src/nunit.analyzers.tests/ClassicModelAssertUsage/ClassicModelAssertUsageAnalyzerTests.cs:97]: I think we should also add
```
                Assert.That(diagnosticIds, Contains.Item(AnalyzerIdentifiers.IsAssignableFromUsage),
                    $"{AnalyzerIdentifiers.IsAssignableFromUsage} is missing.");
                Assert.That(diagnosticIds, Contains.Item(AnalyzerIdentifiers.IsNotAssignableFromUsage),
                    $"{AnalyzerIdentifiers.IsNotAssignableFromUsage} is missing.");
```

File: src/nunit.analyzers.tests/ClassicModelAssertUsage/ClassicModelAssertUsageAnalyzerTests.cs [hunk 2]
```diff
@@ -569,6 +573,38 @@ public void Test()
...
+            var expectedDiagnostic = ExpectedDiagnostic.Create(AnalyzerIdentifiers.NegativeUsage);
+
+            var testCode = TestUtility.WrapClassInNamespaceAndAddUsing(@"
+    public sealed class AnalyzeWhenNegativeIsUsed
+    {
+        public void Test()
+        {
+            ↓ClassicAssert.Negative(2);
+        }
     }");
```
Comment (mikkelbu) [inline src/nunit.analyzers.tests/ClassicModelAssertUsage/ClassicModelAssertUsageAnalyzerTests.cs:608]: Should we also add a method for `IsAssignableFromUsage` and `IsNotAssignableFromUsage`

File: src/nunit.analyzers.tests/ClassicModelAssertUsage/IsAssignableFromClassicModelAssertUsageCodeFixTests.cs [hunk 3]
```diff
@@ -0,0 +1,440 @@
...
+            Assert.That(
+                actual,
+                Is.AssignableFrom(expected),
+                ""message"");
+        }");
+            RoslynAssert.CodeFix(analyzer, fix, expectedDiagnostic, code, fixedCode, fixTitle: ClassicModelAssertUsageCodeFix.TransformToConstraintModelDescription);
+        }
+
+        [Test]
+        public void CodeFixMaintainsReasonableTriviaWithNewLineClosingParen()
```
Comment (mikkelbu) [inline src/nunit.analyzers.tests/ClassicModelAssertUsage/IsAssignableFromClassicModelAssertUsageCodeFixTests.cs:334]: I diffed `IsAssignableFromClassicModelAssertUsageCodeFixTests.cs` and `IsNotAssignableFromClassicModelAssertUsageCodeFixTests.cs` and in the other file this test is parametrized with `[Values] bool hasMessage`. Is there a reason that this one different?

File: src/nunit.analyzers.tests/ClassicModelAssertUsage/IsAssignableFromClassicModelAssertUsageCodeFixTests.cs [hunk 4]
```diff
@@ -0,0 +1,440 @@
...
+
+            Assert.That(
+                actual,
+                Is.AssignableFrom<int>(){commaAndMessage});
+        }}");
+            RoslynAssert.CodeFix(analyzer, fix, expectedDiagnostic, code, fixedCode, fixTitle: ClassicModelAssertUsageCodeFix.TransformToConstraintModelDescription);
+        }
+
+        [Test]
+        public void CodeFixForGenericMaintainsReasonableTriviaWithNewLineClosingParen([Values] bool hasMessage)
```
Comment (mikkelbu) [inline src/nunit.analyzers.tests/ClassicModelAssertUsage/IsAssignableFromClassicModelAssertUsageCodeFixTests.cs:390]: A similar test does not exists in `IsNotAssignableFromClassicModelAssertUsageCodeFixTests.cs`. Is there a reason for this?

File: documentation/index.md [hunk 5]
```diff
@@ -109,6 +109,10 @@ Rules which improve assertions in the test code.
 | [NUnit2048](https://github.com/nunit/nunit.analyzers/tree/master/documentation/NUnit2048.md) | Consider using Assert.That(...) instead of StringAssert(...) | :white_check_mark: | :warning: | :white_check_mark: |
 | [NUnit2049](https://github.com/nunit/nunit.analyzers/tree/master/documentation/NUnit2049.md) | Consider using Assert.That(...) instead of CollectionAssert(...) | :white_check_mark: | :warning: | :white_check_mark: |
 | [NUnit2050](https://github.com/nunit/nunit.analyzers/tree/master/documentation/NUnit2050.md) | NUnit 4 no longer supports string.Format specification | :white_check_mark: | :exclamation: | :white_check_mark: |
+| [NUnit2051](https://github.com/nunit/nunit.analyzers/tree/master/documentation/NUnit2051.md) | Consider using Assert.That(expr, Is.Positive) instead of ClassicAssert.Positive(expr) | :white_check_mark: | :information_source: | :white_check_mark: |
```
Comment (mikkelbu) [inline documentation/index.md:112]: I think we also add the documentation files to the solution file

**Review by mikkelbu — APPROVED** [review]
Great work @manfred-brands. I've added some minor comments/questions, but I'm happy to merge this when you are.

File: src/nunit.analyzers.tests/ClassicModelAssertUsage/ClassicModelAssertUsageAnalyzerTests.cs (hunk 1 above)
Comment (manfred-brands) [reply src/nunit.analyzers.tests/ClassicModelAssertUsage/ClassicModelAssertUsageAnalyzerTests.cs:97]: Done.

File: src/nunit.analyzers.tests/ClassicModelAssertUsage/ClassicModelAssertUsageAnalyzerTests.cs (hunk 2 above)
Comment (manfred-brands) [reply src/nunit.analyzers.tests/ClassicModelAssertUsage/ClassicModelAssertUsageAnalyzerTests.cs:608]: Thanks yes, I must have been rushing it that I forgot. Added.

File: documentation/index.md (hunk 5 above)
Comment (manfred-brands) [reply documentation/index.md:112]: Also forgotten.

File: src/nunit.analyzers.tests/ClassicModelAssertUsage/IsAssignableFromClassicModelAssertUsageCodeFixTests.cs (hunk 3 above)
Comment (manfred-brands) [reply src/nunit.analyzers.tests/ClassicModelAssertUsage/IsAssignableFromClassicModelAssertUsageCodeFixTests.cs:334]: The two files were copied from `IsInstanceOf` and `IsNotInstanceOf` respectively.
Those have the same differences.
I have aligned all 4 now.

File: src/nunit.analyzers.tests/ClassicModelAssertUsage/IsAssignableFromClassicModelAssertUsageCodeFixTests.cs (hunk 4 above)
Comment (manfred-brands) [reply src/nunit.analyzers.tests/ClassicModelAssertUsage/IsAssignableFromClassicModelAssertUsageCodeFixTests.cs:390]: An oversight when I reviewed the PR where the _trivia_ was adjusted. All changed looked fine, but I forgot to check for consistency.
Due to the copying I copied the inconsistency.
Code synced to have same tests for all 4 fixtures.

**Comment by manfred-brands** [comment]
Thanks @mikkelbu for your review and my apologies for some obvious rushed changes.

**Comment by mikkelbu** [comment]
There is absolutely no need to apologise - my comments were the smallest of nit-picks and your contributions the last couple of years are the primary reason that this project has moved forward. So I'm just happy for all of the work that you have put into this repo.

I'll see if I can get a release out tonight.

**Pull request merged**